src/java/voldemort/server/VoldemortServer.java

```diff
@@ -520,24 +524,84 @@ public void goOffline() {
+            return ReadOnlyStoreStatusValidation.NO_STORES_DISABLED;
+        } else {
+            List<String> storesWithDisabledVersions = Lists.newArrayList();
+            for (StorageEngine storageEngine : storageEngines) {
+                StoreVersionManager storeVersionManager = (StoreVersionManager)
+                        storageEngine.getCapability(StoreCapabilityType.DISABLE_STORE_VERSION);
+                if (storeVersionManager.hasAnyDisabledVersion()) {
+                    storesWithDisabledVersions.add(storageEngine.getName());
                 }
             }
-        }
 
-        createOnlineServices();
-        startOnlineServices();
+            if (storesWithDisabledVersions.isEmpty()) {
+                if (voldemortConfig.getHighAvailabilityStateAutoCleanUp()) {
+                    logger.info(VoldemortConfig.PUSH_HA_STATE_AUTO_CLEANUP +
+                                "=true, so the server will attempt to delete the HA state for this node, if any.");
+                    FailedFetchLock failedFetchLock = null;
+                    try {
+                        failedFetchLock = FailedFetchLock.getLock(getVoldemortConfig(), new Props());
+                        failedFetchLock.removeObsoleteStateForNode(getVoldemortConfig().getNodeId());
+                        logger.info("Successfully ensured that the BnP HA shared state is cleared for this node.");
+                    } catch (ClassNotFoundException e) {
+                        logger.error("Failed to find FailedFetchLock class!", e);
+                        return ReadOnlyStoreStatusValidation.UNKNOWN_HA_STATE;
+                    } catch (Exception e) {
+                        logger.error("Exception while trying to remove obsolete HA state!", e);
+                        return ReadOnlyStoreStatusValidation.UNKNOWN_HA_STATE;
+                    } finally {
+                        IOUtils.closeQuietly(failedFetchLock);
+                    }
+                }
+
+                logger.info("No Read-Only stores are disabled. Going online as planned.");
+                return ReadOnlyStoreStatusValidation.NO_STORES_DISABLED;
+            } else {
+                // OMG, there are disabled stores!
+                StringBuilder stringBuilder = new StringBuilder();
+                stringBuilder.append("Cannot go online, because the following Read-Only stores have some disabled version(s): ");
+                boolean firstItem = true;
+                for (String storeName: storesWithDisabledVersions) {
+                    if (firstItem) {
+                        firstItem = false;
+                    } else {
+                        stringBuilder.append(", ");
+                    }
+                    stringBuilder.append(storeName);
+
+                }
+                logger.warn(stringBuilder.toString());
+                return ReadOnlyStoreStatusValidation.SOME_STORES_DISABLED;
```

Comment:
I would prefer to throw an exception with message string you built here. It is hard for a SRE to hop on the box and grep for log message to find the error.